### PR TITLE
feat: [MR-720] Handle `Refund` in streams

### PR DESCRIPTION
**Pull request by alin-at-dfinity**
Replace `RequestOrResponse` with `StreamMessage` in streams, introducing the new `Refund` message. Fully handle anonymous refunds in streams (including crediting cycles, rejecting and rerouting). Define a new certification version (`V22`) to cover all this.

### CODE REVIEW

File: rs/canonical_state/src/encoding/types.rs [hunk 1]
```diff
@@ -377,42 +388,59 @@ pub(crate) fn try_from_deltas(
...
                 response: None,
+                refund: None,
             },
             Response(response) => Self {
                 request: None,
                 response: Some((response.as_ref(), certification_version).into()),
+                refund: None,
             },
+            Refund(refund) => {
+                assert!(certification_version >= CertificationVersion::V22);
```
Comment (stiegerc) [inline rs/canonical_state/src/encoding/types.rs]: Nit: this `assert` is also in the conversion of `Refund` right below.

File: rs/canonical_state/tests/size_limit_visitor.rs [hunk 2]
```diff
@@ -130,9 +130,9 @@ fn compute_message_sizes(state: &ReplicatedState, begin: u64, end: u64) -> usize
     // Sanity check MessageSpyVisitor.
     if let (Some(tbegin), Some(tend)) = (tbegin, tend) {
         assert_eq!((begin, end), (tbegin, tend));
-        // Messages should be at least 35 bytes.
+        // Messages should be at least 25 bytes.
         assert!(
-            size as u64 > (end - begin) * 35,
+            size as u64 > (end - begin) * 25,
```
Comment (stiegerc) [inline rs/canonical_state/tests/size_limit_visitor.rs:135]: Why this change?

File: rs/test_utilities/state/src/lib.rs [hunk 3]
```diff
@@ -924,7 +924,8 @@ prop_compose! {
     )(
         msg_start in msg_start_range,
         msgs in prop::collection::vec(
-            arbitrary::request_or_response_with_config(true),
+            // FIXME
+            arbitrary::stream_message_with_config(false),
```
Comment (stiegerc) [inline rs/test_utilities/state/src/lib.rs]: Don't forget.

File: rs/messaging/src/routing/stream_handler.rs [hunk 4]
```diff
@@ -548,26 +549,26 @@ impl StreamHandlerImpl {
...
-    /// - If the message is a response, it is rerouted according to the routing table into the
-    ///   correspoding stream.
+    /// - If the message is a response or refund, it is rerouted according to the routing table
+    ///   into the appropriate stream.
     ///
     /// Error cases:
-    /// - A response with a `RejectReason` other than `CanisterMigrating`: guaranteed response
-    ///   delivery requires that the response be rerouted; an error counter is incremented.
+    /// - A response or refund with a `RejectReason` other than `CanisterMigrating`: guaranteed
+    ///   response delivery requires responses to be rerouted; an error counter is incremented.
```
Comment (stiegerc) [inline rs/messaging/src/routing/stream_handler.rs]: This makes it sound like rerouting with bad reasons is for responses only. But looking at the code, it's for refunds also right?
I think the comment further down
```
// Refunds are treated the same as responses for rerouting purposes.
```
should be mentioned already here.

Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler.rs]: But since refunds are not absolutely guaranteed, we could also drop them in this case right? Since this can only happen through a bug or maybe corrupted encoding/decoding or something.

**Review by maksymar — APPROVED** [review]

File: rs/canonical_state/src/encoding/types.rs (hunk 1 above)
Comment (alin-at-dfinity) [reply rs/canonical_state/src/encoding/types.rs]: Removed.

File: rs/canonical_state/tests/size_limit_visitor.rs (hunk 2 above)
Comment (alin-at-dfinity) [reply rs/canonical_state/tests/size_limit_visitor.rs:135]: Refunds are smaller than other messages, they only have the one canister ID. So this assert was failing.

And rather than try to do something exceedingly smart (like saying 25 bytes per `Refund`, 35 bytes for everything else) for something that's essentially a sanity check, I just lowered the expected size.

File: rs/test_utilities/state/src/lib.rs (hunk 3 above)
Comment (alin-at-dfinity) [reply rs/test_utilities/state/src/lib.rs]: Removed, thanks for the catch.

This was here because I initially assumed that for `StateManager` tests it would always be OK to have `Refunds` mixed into the streams. However, some of these tests encode said streams into their canonical representation, so they failed to encode `Refunds` with the current certification version (`V21`). I did test it with the current certification version bumped to `V22` and this set to `true` and it works. But testing with `true` will just have to wait until we actually bump to `V22` on `master`.

File: rs/messaging/src/routing/stream_handler.rs [hunk 5]
```diff
@@ -991,6 +1005,67 @@ impl StreamHandlerImpl {
...
+    ) -> Cycles {
+        // Subnet that should have received the message according to the routing table.
+        let receiver_host_subnet = state
+            .metadata
+            .network_topology
+            .route(refund.recipient().get());
+
+        match receiver_host_subnet {
+            // Matching receiver subnet, try crediting the cycles.
+            Some(host_subnet) if host_subnet == self.subnet_id => {
```
Comment (stiegerc) [inline rs/messaging/src/routing/stream_handler.rs:1026]:
```suggestion
            Some(self.subnet_id) => {
```
personal preference.

File: rs/replicated_state/src/metadata_state/tests.rs [hunk 6]
```diff
@@ -1490,6 +1490,40 @@ fn stream_discard_messages_before_returns_expected_messages() {
     assert_eq!(rejected_messages, expected_rejected_messages);
 }
 
+#[test]
+fn stream_discard_messages_before_returns_expected_refunds() {
```
Comment (stiegerc) [inline rs/replicated_state/src/metadata_state/tests.rs:1494]: I kinda prefer one test for all kinds of messages. But for that we'd have to update the testing framework here. Maybe for a cleanup some day...

**Review by stiegerc — COMMENTED** [review]
I didn't see any obvious problems, but I'd like to read the design document first and do another pass first just to make sure I'm up to date on everything.

File: rs/canonical_state/tests/size_limit_visitor.rs (hunk 2 above)
Comment (stiegerc) [reply rs/canonical_state/tests/size_limit_visitor.rs:135]: Thanks for clarifying.

File: rs/messaging/src/routing/stream_handler.rs (hunk 4 above)
Comment (alin-at-dfinity) [reply rs/messaging/src/routing/stream_handler.rs]: Actually, we also want to guarantee delivery of refunds. I've reworded it to say "guaranteed delivery requires rerouting", no longer mentioning responses explicitly.

File: rs/messaging/src/routing/stream_handler.rs (hunk 5 above)
Comment (alin-at-dfinity) [reply rs/messaging/src/routing/stream_handler.rs:1026]: I also thought of that at some point. Unfortunately Rust syntax does not allow that. It has to be a pattern.

File: rs/replicated_state/src/metadata_state/tests.rs (hunk 6 above)
Comment (alin-at-dfinity) [reply rs/replicated_state/src/metadata_state/tests.rs:1494]: I did consider updating one of the existing test cases by manually pushing requests, responses and refunds. But leaving aside the verbose setup, the test case itself would have become too large and difficult to follow.

If anything, I'd rather have smaller, focused test cases like this; plus a smoke test showing that a mix of message types is also inducted correctly (and e.g. there's no mix-up in reject signal indices).

Comment (stiegerc) [reply rs/replicated_state/src/metadata_state/tests.rs:1494]: Right, it's just that I made these stream generator things for the stream handler tests that also use a config. So we could do something like that here, i.e. make the `MessageConfig` more than just a number of messages. Maybe in a refactor some day. For now this seems sufficient.

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 7]
```diff
@@ -1419,6 +1425,75 @@ fn garbage_collect_local_state_with_reject_signals_for_request_from_migrating_ca
...
+/// when the recipient is on this subnet; and in its rerouting otherwise.
+#[test]
+fn garbage_collect_local_state_with_reject_signals_for_misrouted_refunds() {
+    with_test_setup(
+        // An outgoing stream with two refund, @21 and @22.
+        btreemap![REMOTE_SUBNET => StreamConfig {
+            begin: 21,
+            messages: vec![Refund(*LOCAL_CANISTER), Refund(*OTHER_REMOTE_CANISTER)],
+            ..StreamConfig::default()
+        }],
```
Comment (stiegerc) [inline rs/messaging/src/routing/stream_handler/tests.rs:1438]: I'm a bit confused by this setup. Why are we testing a refund addressed to a local canister in a stream to a remote subnet? I suppose it could happen if the canister was just moved here from the remote subnet, but how is it then different from the other refund that also gets rerouted?

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 8]
```diff
@@ -2576,6 +2652,111 @@ fn system_subnet_induct_stream_slices_with_subnet_message_memory_limit() {
...
+                ],
+                ..StreamConfig::default()
+            });
+            expected_state.with_streams(btreemap![REMOTE_SUBNET => expected_stream]);
+
+            // Refund to `LOCAL_CANISTER` (@43) was applied.
+            let StreamMessage::Refund(refund43) = message_in_slice(slices.get(&REMOTE_SUBNET), 43)
+            else {
+                panic!("Expected a refund message");
+            };
```
Comment (stiegerc) [inline rs/messaging/src/routing/stream_handler/tests.rs]: We have functions like `response_in_slice` for exactly this use case. I suggest we add a `refund_in_slice` analogously.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 7 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs:1438]: Also we should be able to extend an existing test with a refund no?

File: rs/messaging/src/routing/stream_handler.rs (hunk 5 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler.rs:1026]: Probably the same reason you can't match into `Arcs`. In that case nevermind.

File: rs/messaging/src/routing/stream_handler.rs (hunk 4 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler.rs]: Read the design document again, makes more sense now.

File: rs/replicated_state/src/metadata_state/tests.rs (hunk 6 above)
Comment (stiegerc) [reply rs/replicated_state/src/metadata_state/tests.rs:1494]: Or use the same thing, which is what I intended to do at the time, but didn't for some reason that escapes me now :-/

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 8 above)
Comment (alin-at-dfinity) [reply rs/messaging/src/routing/stream_handler/tests.rs]: Done.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 7 above)
Comment (alin-at-dfinity) [reply rs/messaging/src/routing/stream_handler/tests.rs:1438]: The test validates the behavior of `garbage_collect_local_state()` with reject signals for refunds: the refund is correctly rerouted, whether the recipient is remote or local. A local recipient is an edge case that (in the absence of any assumptions regarding the implementation) may well result in the refund being dropped or otherwise mishandled.

Regarding extending an existing test, (i) this is a test for `garbage_collect_local_state`'s handling of reject signals; and (ii) existing tests for this behavior are specific to the type of message (e.g. `garbage_collect_local_state_with_reject_signals_for_request_from_migrating_canister()` above).

File: rs/replicated_state/src/metadata_state/tests.rs (hunk 6 above)
Comment (alin-at-dfinity) [reply rs/replicated_state/src/metadata_state/tests.rs:1494]: As said, I tend to prefer small, focused test cases where possible (although I did write quite a few humongous, convoluted MR test cases, mostly by necessity). So while I wouldn't mind a test with a more concise setup, I would still keep it as a separate test.

File: rs/types/types_test_utils/src/arbitrary.rs [hunk 9]
```diff
@@ -212,14 +212,24 @@ prop_compose! {
     }
 }
 
-/// Produces an arbitrary [`RequestOrResponse`], with the `deadline` field
-/// populated or not.
-pub fn request_or_response_with_config(
-    populate_deadline: bool,
-) -> impl Strategy<Value = RequestOrResponse> {
+prop_compose! {
```
Comment (berestovskyy) [inline rs/types/types_test_utils/src/arbitrary.rs:215]: nit: the code inside the `prop_compose` macro can't be auto-formatted with rustfmt, so I'd go with explicit `-> impl Strategy<...>`

**Review by berestovskyy — APPROVED** [review]

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 7 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs:1438]: We have a case like this one [for requests](https://sourcegraph.com/github.com/dfinity/ic@ce3ee70c263d5dd75fb5798068c3c9276b76336b/-/blob/rs/messaging/src/routing/stream_handler/tests.rs?L1292), but we specifically call it misrouted. I suppose it's a valid edge case, but a refund to a canister on this subnet in any other stream than the loopback stream is misrouted, except perhaps for the case I mentioned above.

File: rs/types/types_test_utils/src/arbitrary.rs (hunk 9 above)
Comment (alin-at-dfinity) [reply rs/types/types_test_utils/src/arbitrary.rs:215]: I see your point. However, all similar methods (except the ones consisting of a single `prop_oneof!`) are written as a `prop_compose!`.

I'd rather stay consistent (this is usually one of the first recommendations of any code style guide) and do it all at once if at all.
